semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/Javacp.scala

```diff
@@ -0,0 +1,402 @@
+      uri = uri,
+      symbols = symbols
+    )
+  }
+
+  val javaLanguage = Some(s.Language("Java"))
+
+  def fromJavaTypeSignature(sig: JavaTypeSignature)(implicit scopes: TypeVariableScopes): s.Type =
+    sig match {
+      case ClassTypeSignature(SimpleClassTypeSignature(identifier, targs), suffix) =>
+        val prefix = styperef(ssym(identifier), targs.toType)
+        suffix.foldLeft(prefix) {
+          case (accum, s: ClassTypeSignatureSuffix) =>
+            styperef(
+              prefix = Some(accum),
+              symbol = ssym(s.simpleClassTypeSignature.identifier),
+              args = s.simpleClassTypeSignature.typeArguments.toType
+            )
+        }
+      case TypeVariableSignature(name) =>
+        styperef(scopes.resolve(name))
+      case t: BaseType =>
+        styperef("_root_.scala." + t.name + "#")
+      case ArrayTypeSignature(tpe) =>
+        sarray(tpe.toType)
+    }
+
+  case class TypeParameterInfo(value: TypeParameter, symbol: String)
+  def addTypeParameters(
+      typeParameters: TypeParameters,
+      ownerSymbol: String,
+      scopes: TypeVariableScopes): List[s.SymbolInformation] = {
+    val infos = typeParameters.all.map { typeParameter: TypeParameter =>
+      val symbol = ownerSymbol + "[" + typeParameter.identifier + "]"
+      // need to register all tparams before computing rhs types
+      scopes.enterBinding(ownerSymbol, typeParameter.identifier, symbol)
+      TypeParameterInfo(typeParameter, symbol)
+    }
+    infos.map(info => addTypeParameter(info, ownerSymbol, scopes))
+  }
+  def addTypeParameter(
+      typeParameter: TypeParameterInfo,
+      ownerSymbol: String,
+      scopes: TypeVariableScopes): s.SymbolInformation = {
+    val typeParameters = typeParameter.value.upperBounds.map(fromJavaTypeSignature(_)(scopes))
+    val upperBounds = typeParameters match {
+      case upperBound :: Nil =>
+        upperBound
+      case _ =>
+        s.Type(
+          s.Type.Tag.STRUCTURAL_TYPE,
+          structuralType = Some(s.StructuralType(parents = typeParameters))
+        )
+    }
+    val tpe = s.Type(
+      s.Type.Tag.TYPE_TYPE,
+      typeType = Some(s.TypeType(upperBound = Some(upperBounds)))
+    )
+
+    s.SymbolInformation(
+      symbol = typeParameter.symbol,
+      language = javaLanguage,
+      kind = k.TYPE_PARAMETER,
+      name = typeParameter.value.identifier,
+      tpe = Some(tpe),
+      owner = ownerSymbol
+    )
+  }
+
+  def methodDescriptor(signature: MethodSignature): String =
+    signature.params.iterator
+      .map {
+        case t: BaseType => t.name
+        case t: ClassTypeSignature => sname(t.simpleClassTypeSignature.identifier)
+        case t: TypeVariableSignature => t.identifier
+        case _: ArrayTypeSignature => "Array"
+      }
+      .mkString(",")
+
+  case class MethodInfo(node: MethodNode, descriptor: String, signature: MethodSignature)
+
+  def ssymbols(node: ClassNode, scopes: TypeVariableScopes): Seq[s.SymbolInformation] = {
```

Comment:
I think this should be called `sinfos` to communicate that it returns a list of `s.SymbolInformation`, not `s.Symbol`.

Reply:
Looks done.